examples/vision/pointnet.py

```diff
@@ -0,0 +1,295 @@
+mesh.show()
+
+"""
+To convert a mesh file to a point cloud we first need to sample points on the mesh
+surface. `.sample()` performs a unifrom random sampling. Here we sample at 2048 locations
+and visualize in `matplotlib`.
+"""
+
+points = mesh.sample(2048)
+trimesh.points.plot_points(points)
+
+"""
+To generate a `tf.data.Dataset()` we need to first parse through the ModelNet data
+folders. Each mesh is loaded and sampled into a point cloud before being added to a
+standard python list and converted to a `numpy` array. We also store the current
+enumerate index value as the object label.
+"""
+
+
+def parse_dataset(n_pts=2048):
+
+    train_pts = []
+    train_labels = []
+    test_pts = []
+    test_labels = []
+
+    folders = glob.glob("ModelNet10/[!README]*")
+
+    for i, folder in enumerate(folders):
+
+        print("processing: %s" % folder)
+        train_files = glob.glob(folder + "/train/*")
+        test_files = glob.glob(folder + "/test/*")
+
+        for f in train_files:
+            train_pts.append(trimesh.load(f).sample(n_pts))
+            train_labels.append(i)
+
+        for f in test_files:
+            test_pts.append(trimesh.load(f).sample(n_pts))
+            test_labels.append(i)
+
+    return (
+        np.array(train_pts),
+        np.array(test_pts),
+        np.array(train_labels),
+        np.array(test_labels),
+    )
+
+
+"""
+Set the number of points to sample and batch size and parse the dataset. This can take
+~5minutes to complete.
+"""
+
+N_PTS = 2048
+NUM_CLASSES = 10
+BATCH_SIZE = 32
+
+X_train, X_test, y_train, y_test = parse_dataset(N_PTS)
+
+"""
+Our data can now be read into a `tf.data.Dataset()` object. We set the shuffle buffer
+size to the entire size of the dataset as prior to this the data is ordered by class.
+Data augmentation is important when working with point cloud data. We create a
+augmentation function to jitter, shuffle and rotate the train dataset.
+"""
+
+
+def augment(pts, label):
+
+    # jitter points
+    pts += tf.random.uniform(pts.shape, -0.005, 0.005, dtype=tf.float64)
+    # shuffle points
+    pts = tf.random.shuffle(pts)
+    # random rotation about the z axis
+    theta = np.random.uniform(-np.pi, np.pi)
+    c = np.cos(theta)
+    s = np.sin(theta)
+    R = tf.constant([[c, -s, 0.0], [s, c, 0.0], [0.0, 0.0, 1.0]], tf.float64)
+    pts = tf.einsum("...ij,...kj->...ki", R, pts)
+
+    return pts, label
+
+
+train_dataset = tf.data.Dataset.from_tensor_slices((X_train, y_train))
+test_dataset = tf.data.Dataset.from_tensor_slices((X_test, y_test))
+
+train_dataset = train_dataset.shuffle(len(X_train)).map(augment).batch(BATCH_SIZE)
+test_dataset = test_dataset.shuffle(len(X_test)).batch(BATCH_SIZE)
+
+"""
+### Build a model
+"""
+
+"""
+Each convolution and fully-connected layer (with exception for end layers) consits of
+Convolution / Dense -> Batch Normalization -> ReLU Activation. We can make simple
+sub-classed layer to perform all these as a single layer.
```

Comment:
You don't need to, you can take a fully functional approach, which is more concise and does not create additional deserialization overhead

Reply:
I was trying to make layer declaration / calling consistent so now it would be for example:
x = conv_bn(x, n)
x = layers.GlobalMaxPooling1D()(x)

Would it also be better to change the T-Net to be fully functional in this style, or better left as their own model objects?

Reply:
A function that returns a `Model` instance is a perfectly fine pattern, no problem there.